MixinServices/MixinServices/Services/Crypto/TIP/TIP.swift

```diff
@@ -0,0 +1,459 @@
+            + pinIterator().data(endianness: .little)
+        let based = try AESCryptor.encrypt(pinData, with: pinToken).base64RawURLEncodedString()
+        return based
+    }
+    
+    static func encryptTIPPIN(pin: String, pinToken: Data, target: Data) async throws -> String {
+        guard let pinData = pin.data(using: .utf8) else {
+            throw Error.invalidPIN
+        }
+        let tipPriv: Data
+        if let savedTIPPriv = AppGroupKeychain.tipPriv {
+            let aesKey = try await getAESKey(pinData: pinData, pinToken: pinToken)
+            guard let tipPrivKey = SHA3_256.hash(data: aesKey + pinData) else {
+                throw Error.unableToHashTIPPrivKey
+            }
+            tipPriv = try AESCryptor.decrypt(savedTIPPriv, with: tipPrivKey)
+        } else {
+            tipPriv = try await createTIPPriv(pin: pin,
+                                              failedSigners: [],
+                                              legacyPIN: nil,
+                                              forRecover: true,
+                                              progressHandler: nil)
+        }
+        return try encryptTIPPIN(tipPriv: tipPriv, target: target)
+    }
+    
+    static func pinIterator() -> UInt64 {
+        var iterator: UInt64 = 0
+        PropertiesDAO.shared.updateValue(forKey: .iterator, type: UInt64.self) { databaseValue in
+            let userDefaultsValue = AppGroupUserDefaults.Crypto.iterator
+            if let databaseValue = databaseValue {
+                if databaseValue != userDefaultsValue {
+                    Logger.general.warn(category: "TIP", message: "database: \(databaseValue), defaults: \(userDefaultsValue)")
+                }
+                iterator = max(databaseValue, userDefaultsValue)
+            } else {
+                iterator = userDefaultsValue
+                Logger.general.info(category: "TIP", message: "Iterator initialized to \(userDefaultsValue)")
+            }
+            let nextIterator = iterator + 1
+            AppGroupUserDefaults.Crypto.iterator = nextIterator
+            return nextIterator
+        }
+        Logger.general.info(category: "TIP", message: "Encrypt with it: \(iterator)")
+        return iterator
+    }
+    
+    static func ephemeralSeed(pinToken: Data) async throws -> Data {
+        if let seed = AppGroupKeychain.ephemeralSeed {
+            Logger.general.debug(category: "TIP", message: "Using saved ephemeral: \(seed.hexEncodedString())")
+            return seed
+        } else {
+            let ephemerals = try await TIPAPI.ephemerals()
+            if let newest = ephemerals.max(by: { $0.createdAt > $1.createdAt }), let decoded = Data(base64URLEncoded: newest.seed) {
+                let seed = try AESCryptor.decrypt(decoded, with: pinToken)
+                try await TIPAPI.updateEphemeral(base64URLEncoded: newest.seed)
+                AppGroupKeychain.ephemeralSeed = seed
```

Comment:
same as [it](https://github.com/MixinNetwork/ios-app/pull/1096#pullrequestreview-1108582327)